flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderInformationRegisterTest.java

```diff
@@ -0,0 +1,244 @@
+        final String contenderID = "contender-id";
+        final LeaderInformation leaderInformation =
+                LeaderInformation.known(UUID.randomUUID(), "address");
+
+        final LeaderInformationRegister testInstance =
+                LeaderInformationRegister.of(contenderID, leaderInformation);
+        assertThat(testInstance.getRegisteredContenderIDs()).containsExactly(contenderID);
+        assertThat(testInstance.forContenderID(contenderID)).hasValue(leaderInformation);
+    }
+
+    @Test
+    void testOfWithEmptyLeaderInformation() {
+        final String contenderID = "contender-id";
+        final LeaderInformationRegister testInstance =
+                LeaderInformationRegister.of(contenderID, LeaderInformation.empty());
+
+        assertThat(testInstance.getRegisteredContenderIDs()).isEmpty();
+        assertThat(testInstance.forContenderID(contenderID)).isNotPresent();
+    }
+
+    @Test
+    void testMerge() {
+        final String contenderID = "contender-id";
+        final LeaderInformation leaderInformation =
+                LeaderInformation.known(UUID.randomUUID(), "address");
+
+        final String newContenderID = "new-contender-id";
+        final LeaderInformation newLeaderInformation =
+                LeaderInformation.known(UUID.randomUUID(), "new-address");
+
+        final LeaderInformationRegister initialRegister =
+                LeaderInformationRegister.of(contenderID, leaderInformation);
+
+        final LeaderInformationRegister newRegister =
+                LeaderInformationRegister.merge(
+                        initialRegister, newContenderID, newLeaderInformation);
+
+        assertThat(newRegister).isNotSameAs(initialRegister);
+        assertThat(newRegister.getRegisteredContenderIDs())
+                .containsExactlyInAnyOrder(contenderID, newContenderID);
+        assertThat(newRegister.forContenderID(contenderID)).hasValue(leaderInformation);
+        assertThat(newRegister.forContenderID(newContenderID)).hasValue(newLeaderInformation);
+    }
+
+    @Test
+    void testMergeEmptyLeaderInformation() {
+        final String contenderID = "contender-id";
+        final LeaderInformation leaderInformation =
+                LeaderInformation.known(UUID.randomUUID(), "address");
+
+        final String newContenderID = "new-contender-id";
+
+        final LeaderInformationRegister initialRegister =
+                LeaderInformationRegister.of(contenderID, leaderInformation);
+
+        final LeaderInformationRegister newRegister =
+                LeaderInformationRegister.merge(
+                        initialRegister, newContenderID, LeaderInformation.empty());
+
+        assertThat(newRegister).isNotSameAs(initialRegister);
+        assertThat(newRegister.getRegisteredContenderIDs()).containsExactly(contenderID);
+        assertThat(newRegister.forContenderID(newContenderID)).isNotPresent();
+    }
+
+    @Test
+    void testMergeEmptyLeaderInformationForExistingContenderID() {
```

Comment:
for completeness a test where we overwrite existing leader information with new (non-empty) leader information would be nice.